sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts

```diff
@@ -673,17 +707,198 @@ export class EducationProgramOfferingService extends RecordDataModelService<Educ
+
+    //Populate the status accordingly and get impacted applications
+    // when the offering change is approved.
+    if (offeringStatus === OfferingStatus.Approved) {
+      requestedOffering.offeringStatus = OfferingStatus.Approved;
+      precedingOffering.offeringStatus = OfferingStatus.ChangeOverwritten;
+      applications = await this.getApplicationsToSubmitReassessment(offeringId);
+
+      for (const application of applications) {
+        application.currentAssessment = {
+          application: { id: application.id } as Application,
+          triggerType: AssessmentTriggerType.OfferingChange,
+          offering: { id: offeringId } as EducationProgramOffering,
+          creator: auditUser,
+          createdAt: currentDate,
+          submittedBy: auditUser,
+          submittedDate: currentDate,
+        } as StudentAssessment;
+        application.modifier = auditUser;
+        application.updatedAt = currentDate;
+      }
+    } else {
+      requestedOffering.offeringStatus = OfferingStatus.ChangeDeclined;
+      precedingOffering.offeringStatus = OfferingStatus.Approved;
+    }
+
+    await this.dataSource.transaction(async (transactionalEntityManager) => {
+      // create the note for assessment.
+      const note = new Note();
+      note.description = assessmentNotes;
+      note.noteType = NoteType.Program;
+      note.creator = auditUser;
+      const noteEntity = await transactionalEntityManager
+        .getRepository(Note)
+        .save(note);
+
+      // update note.
+      requestedOffering.offeringNote = noteEntity;
+
+      // update institution note.
+      await transactionalEntityManager
+        .getRepository(Institution)
+        .createQueryBuilder()
+        .relation(Institution, "notes")
+        .of({
+          id: requestedOffering.institutionLocation.institution.id,
+        } as Institution)
+        .add(noteEntity);
+
+      //Save the requested and preceding offering.
+      await transactionalEntityManager
+        .getRepository(EducationProgramOffering)
+        .save([requestedOffering, precedingOffering]);
+
+      //Save applications with new current assessment on approval
+      if (applications && applications.length > 0) {
+        await transactionalEntityManager
+          .getRepository(Application)
+          .save(applications);
+      }
+    });
+
+    //Once the impacted applications are updated with new current assessment
+    //start the assessment workflows and delete the existing workflow instances.
+    await this.startOfferingChangeAssessments(applications);
+  }
+
+  /**
+   * For an offering change start the assessment workflows for all new assessments
+   * and delete the existing workflow instances of previous assessments.
+   * @param applications applications impacted by offering change.
+   */
+  private async startOfferingChangeAssessments(
+    applications: ApplicationAssessmentSummary[],
+  ): Promise<void> {
```

Comment:
I saw an article on dealing with high cpu root causes in Nodejs.
https://medium.com/geekculture/dealing-with-node-js-high-cpu-in-production-71c432d8bece

Reply:
Thanks for sharing it and I read it. But as per our code we create only a certain number of parallel instances at a time which is almost equal to number of logical CPU cores. And we wait until these parallel instances are complete before we start any new instance. Hence we are not likely to flood our cpu here. IMO.

Reply:
Thanks for providing further info during our call. I will resolve.

Reply:
This is not a high CPU process it is a high IO process 😉 
It will be calling an external API executing IO-intensive operations, not CPU intensive.